crates/rover-client/src/query/subgraph/list.rs

```diff
@@ -0,0 +1,194 @@
+use crate::blocking::StudioClient;
+use crate::RoverClientError;
+use chrono::prelude::*;
+use graphql_client::*;
+
+type Timestamp = String;
+
+#[derive(GraphQLQuery)]
+// The paths are relative to the directory where your `Cargo.toml` is located.
+// Both json and the GraphQL schema language are supported as sources for the schema
+#[graphql(
+    query_path = "src/query/subgraph/list.graphql",
+    schema_path = ".schema/schema.graphql",
+    response_derives = "PartialEq, Debug, Serialize, Deserialize",
+    deprecated = "warn"
+)]
+/// This struct is used to generate the module containing `Variables` and
+/// `ResponseData` structs.
+/// Snake case of this name is the mod name. i.e. list_subgraphs_query
+pub struct ListSubgraphsQuery;
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct SubgraphInfo {
+    pub name: String,
+    pub url: Option<String>, // optional, and may not be a real url
+    pub updated_at: Option<DateTime<Local>>,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct ListDetails {
+    pub subgraphs: Vec<SubgraphInfo>,
+    pub root_url: String,
+    pub graph_name: String,
+}
+
+/// Fetches list of subgraphs for a given graph, returns name & url of each
+pub fn run(
+    variables: list_subgraphs_query::Variables,
+    client: &StudioClient,
+) -> Result<ListDetails, RoverClientError> {
+    let graph_name = variables.graph_id.clone();
+    let response_data = client.post::<ListSubgraphsQuery>(variables)?;
+    let root_url = response_data.frontend_url_root.clone();
+    let subgraphs = get_subgraphs_from_response_data(response_data, &graph_name)?;
+    Ok(ListDetails {
+        subgraphs: format_subgraphs(&subgraphs),
+        root_url,
+        graph_name,
+    })
+}
+
+type RawSubgraphInfo = list_subgraphs_query::ListSubgraphsQueryServiceImplementingServicesOnFederatedImplementingServicesServices;
+fn get_subgraphs_from_response_data(
+    response_data: list_subgraphs_query::ResponseData,
+    graph_name: &str,
+) -> Result<Vec<RawSubgraphInfo>, RoverClientError> {
+    let service_data = match response_data.service {
+        Some(data) => Ok(data),
+        None => Err(RoverClientError::NoService),
+    }?;
+
+    // get list of services
+    let services = match service_data.implementing_services {
+        Some(services) => Ok(services),
+        // this case may be removable in the near future as unreachable, since
```

Comment:
Ohhh can we add a `TODO (api-error)` thing here? This seems like a prime candidate of an API error that needs to change / will change in the future

Reply:
Ah, this _was_ updated, but still needs some api attention. this case should be unreachable now, but can't be removed yet, because that key in the graphql request is technically still nullable. I'll add that note :)